### ISSUE TITLE
Provide official docker image with pdk

### ISSUE THREAD

**Issue by gdubicki** [puppetlabs/pdk#336]
As various CI solution, such as GitLab CI, are using Docker containers to run the jobs inside them, it would be nice if you could provide official Docker image with pdk. Preferably it should be based on Alpine for minimal size.

**Comment by DavidS** [comment]
Hi, we're tracking this in https://tickets.puppetlabs.com/browse/PDK-629 ; see https://github.com/sharumpe/puppet-pdk for some unofficial work in that area.

**Comment by JacobHenner** [comment]
I created https://hub.docker.com/r/jacobhenner/puppet-test/ and https://hub.docker.com/r/jacobhenner/puppet-pdk/ based off of the Docker ruby images. I intend to tag them for alternative base images (e.g. Alpine), perhaps these are useful?

**Comment by zoidbergwill** [comment]
@JacobHenner having images with tagged versions of pdk would be awesome.

**Comment by DavidS** [comment]
There's also https://github.com/ccaum/puppet-dev-in-docker which has some of our tooling, and could use some love.

**Comment by ccaum** [comment]
If anyone is interested in trying it out, I have a newer version of the puppet-dev-tools image here: https://hub.docker.com/r/ccaum/puppet-dev-tools/

It's based on CentOS so it can better support PDK.

**Comment by zoidbergwill** [comment]
Is there a repo for that?

**Comment by ccaum** [comment]
For now the repo is here: https://github.com/ccaum/puppet-dev-in-docker
I'll move it into the puppetlabs namespace once I'm happy with it.

I might split it into two images. One with just PDK, onceover, etc. and another built on it for the CD for PE specific bits. Feedback welcome.

**Comment by zoidbergwill** [comment]
Ahh, had found that repo, but didn't think `ccaum/puppet-dev-tools` was from there for some reason. I was being a potato. That looks awesome.

**Comment by bittner** [comment]
Is it so hard to provide an "official" Docker image, something we pull with [puppet/pdk](https://hub.docker.com/r/puppet/pdk)? This would really help standardize running PDK in CI environments.

A partial solution would be to provide a Docker image from the Vox Pupuli namespace (i.e. [voxpupuli/pdk](https://hub.docker.com/r/voxpupuli/pdk)) as a community effort, but then again PDK isn't really a @voxpupuli project.

**Comment by TJM** [comment]
I have found puppet/puppet-dev-tools, which included PDK and some other stuff. It is what we use in our CI Puppet Validation. I am working on making it a little "smaller" (more efficient). This does not replace the desire to have an "official" puppet/pdk image, but provides a very useful workaround for now.

Tommy

**Comment by glennsarti** [comment]
@rodjek @ccaum As this is now in the puppetlabs space, is this now "official" enough?

**Comment by scotje** [comment]
https://hub.docker.com/r/puppet/pdk is now available!

`latest` tag will get you latest stable release of PDK

`nightly` tag will get you the very latest (which might be the same as `latest` or might be a nightly pre-release, whatever is most recent)

There are also tags for each version (including pre-releases).

Please let us know if you have any issues with the images!

**Comment by scotje** [comment]
Note that there will be some lag time (on the order of an hour) between when something is merged/tagged on the PDK repo and when updated Docker images are available. 

Also availability of nightly builds containing specific recent commits is a best effort thing.

**Comment by bittner** [comment]
The [puppet/pdk image](https://hub.docker.com/r/puppet/pdk) is quite large:

```
$ docker image ls
REPOSITORY        TAG            IMAGE ID            CREATED             SIZE
puppet/pdk        latest         56136175d031        2 days ago          802MB
```

To be sensibly usable for CI it would be good to drastically reduce the image size. Can you use a smaller base image? I'm sure that must be possible.

**Comment by scotje** [comment]
The ubuntu base image layer is 27MB compressed and 63MB uncompressed. The PDK package installation layer is about 667MB uncompressed. I don't think switching to Alpine is going to drastically reduce the size. (Disregarding that we don't build a package for Alpine currently.)

Ideally a CI system should be making use of image caching so that it doesn't have to pull down the whole image (which is about 337MB compressed).

If someone wants to write a gem-installation based Dockerfile that we could offer as another tag or in another repo I'd be happy to review/merge it into the https://github.com/puppetlabs/pdk-docker repo. But it should be in parallel to the current implementation since a Gem-based installation is not going to have all the features of a package-based installation.

**Comment by TJM** [comment]
You beat me to it. I was going to say that I had tried reducing the size for https://hub.docker.com/r/puppet/puppet-dev-tools ... and while I made some progress, PDK is pretty *bulky*. Maybe we need to put some effort into a PDK lite, but that is outside the scope of providing the image, right :)

Thanks to @scotje and to everyone else behind the scenes for providing this.

~tommy

**Comment by ccaum** [comment]
Yeah I promise I tried to get the image as small as possible. @TJM did some good work but ultimately it's down to the Rubies that ship with PDK. I originally built it on Alpine and it was indeed a more reasonable size but didn't install PDK the correct way so it would only be able to run against a single version of Puppet.

**Comment by LongLiveCHIEF** [comment]
> If someone wants to write a gem-installation based Dockerfile that we could offer as another tag or in another repo I'd be happy to review/merge it

I did that already... 2 years ago! :smile: see #342.  

After a docker pull, it comes out to about 430MB.
